### ISSUE TITLE
Interoperability with native APIs

### ISSUE THREAD

**Issue by xibaoning** [trungnt2910/MemoryModule.NET#42]
how get DllImportResolver IntPtr?

```csharp
  NativeLibrary.SetDllImportResolver(typeof(SQLitePCL.SQLite3Provider_e_sqlite3).Assembly, DllImportResolver);

private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath){
 var ass = NativeAssembly.Load(File.ReadAllBtytes("e_sqllite3.dll"));

 var handle = ass.GetType().GetField("_handle", BindingFlags.Instance | BindingFlags.NonPublic)?.GetValue(ass);

return handle+sizeof(IntPtr); ?  

}
```

**Comment by trungnt2910** [comment]
Unfortunately, `MemoryModule` is currently not compatible with the .NET Core/.NET 5+ only `NativeLibrary` and its supporting functions.

Doing so would require diving in the runtime's deep internals, not to mention having to add .NET Core-specific guards (we still want to support .NET Standard).

If you want to get functions, you should use the `GetDelegate` functions. To make it look like the usual `DllImport`s, maybe storing everything in `static readonly` fields that are initialized in a `static` constructor should work?

**Comment by xibaoning** [comment]
private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
 {
     var ass = NativeAssembly.Load(File.ReadAllBytes("e_sqllite3.dll"));
     var funcaddress = ass.GetDelegate<GetIntDelegate>("sqlite3_libversion_number"); //isok
     var version=funcaddress(); //isok

     //other code exception
     System.EntryPointNotFoundException: Unable to find an entry point named 'sqlite3_libversion_number' in DLL 'e_sqlite3'.
 }

**Comment by trungnt2910** [comment]
No, you should not use the built-in `DllImport`.

What you should use looks more like:

```csharp
public static class Sqlite3
{
    public static readonly GetIntDelegate LibversionNumber;

    static Sqlite3()
    {
         var asm = NativeAssembly.Load(File.ReadAllBytes("e_sqllite3.dll"));
         LibversionNumber = asm.GetDelegate<GetIntDelegate>("sqlite3_libversion_number");
    }
}
```

The original `DllImport` would not work (and therefore the official `sqlite` bindings).

It is impossible to hook into `NativeLibrary` itself, since `NativeLibrary` is just a thin wrapper around OS functions (`GetProcAddress`, `dlsym`), hooking into _that_ would require messing much more than just .NET Runtime internals.

**Comment by xibaoning** [comment]
OK, THS




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Trung ***@***.***&gt;; 
发送时间: 2023年12月29日(星期五) 中午12:49
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [trungnt2910/MemoryModule.NET] Interoperability with `System.Runtime.InteropServices.NativeLibrary.SetDllImportResolver` (Issue #42)





 
No, you should not use the built-in DllImport.
 
What you should use looks more like:
 public static class Sqlite3 {     public static readonly GetIntDelegate LibversionNumber;     static Sqlite3()     {          var asm = NativeAssembly.Load(File.ReadAllBytes("e_sqllite3.dll"));          LibversionNumber = asm.GetDelegate<GetIntDelegate&gt;("sqlite3_libversion_number");     } }
 
The original DllImport would not work (and therefore the official sqlite bindings).
 
It is impossible to hook into NativeLibrary itself, since NativeLibrary is just a thin wrapper around OS functions (GetProcAddress, dlsym), hooking into that would require messing much more than just .NET Runtime internals.

**Comment by xibaoning** [comment]
```C#
 private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
 {
	  var sqlitehandle = NativeLibrary.Load("e_sqlite3.dll"); 	  
	  return sqlitehandle; //this worked ok
	  
	  //var ass = NativeAssembly.Load(File.ReadAllBytes("e_sqllite3.dll"));
	  //any way to get right handle from NativeAssembly
 }
```

**Comment by trungnt2910** [comment]
`NativeLibrary.Load` is also just a thin wrapper around the Windows `LoadLibrary` and the POSIX `dlopen`, which requires a file on the disk. You can see the source of .NET for more details: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/nativelibrary.cpp

It is not possible to use these APIs for assemblies solely in-memory. Unless, for example, on Linux, you can `memfd_open` that region, and then use `/proc/self/fd/[whatever FD you just received]`. I am not aware of any similar APIs on Windows, or even other *NIXes though.

TL; DR: 不可能.

**Comment by trungnt2910** [comment]
@xibaoning

It appears that you are using this only on Windows.
On Windows, there are many more powerful `MemoryModule`-style libraries out there that are written in C. One of them is [`MemoryModulePP`](https://github.com/bb107/MemoryModulePP), which supports usage with Win32 APIs like `GetProcAddress`, and therefore also compatible with .NET's `NativeLibrary` on Windows.

An example of bootstrapping to `MemoryModulePP` from `MemoryModule.NET` can be found here:
https://github.com/trungnt2910/MemoryModulePP.NET

In your case, what you should do is:

```csharp
 private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
 {
	  using var memoryModulePP = NativeAssembly.Load(File.ReadAllBytes("MemoryModulePP.dll"));

	  var LdrLoadDllMemoryExW = memoryModulePP.GetDelegate<LdrLoadDllMemoryExWDelegate>("LdrLoadDllMemoryExW");

	  const uint LOAD_FLAGS_PASS_IMAGE_CHECK = 0x40000000;

	  LdrLoadDllMemoryExW(
    	  	  out IntPtr handle,
    	  	  out _,
    	  	  // MemoryModulePP has a weird check that will somehow fail for our binaries.
	  	  LOAD_FLAGS_PASS_IMAGE_CHECK,
    	  	  File.ReadAllBytes("e_sqlite3.dll"),
    	  	  // This misleading parameter must always be 0.
    	  	  0,
    	  	  null,
    	  	  null
	  );

	  return handle;
 }
```

**Comment by xibaoning** [comment]
Worked! You did a great job!

**Comment by trungnt2910** [comment]
Closing this issue for now then. I have put a link to the relevant guides on the README page.
